### PR TITLE
Add IPProto, ConditionFlag and AppID.

### PR DESCRIPTION
**Pull request by maisem**
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: firewall.go [hunk 1]
```diff
@@ -109,6 +109,53 @@ func (id *LayerID) IsZero() bool {
...
+	IPProtoUDP    IPProto = 17
+)
+
+// AppID returns the application ID associated with the provided file.
+func AppID(file string) (string, error) {
+	var a arena
+	defer a.Dispose()
+	fileBytes, _ := toBytesFromString(&a, file)
+	appID := (*fwpByteBlob)(a.Alloc(unsafe.Sizeof(fwpByteBlob{})))
+	if err := fwpmGetAppIdFromFileName0(fileBytes, &appID); err != nil {
```
Comment (danderson) [inline firewall.go:153]: This call isn't right, the function returns allocated memory, and you need to free it with fwpmFreeMemory0. See https://docs.microsoft.com/en-us/windows/win32/api/fwpmu/nf-fwpmu-fwpmgetappidfromfilename0 for how it's used in C++.

Should be something like:

var blob *fwpByteBlob
fwpmGetAppIdFromFileName0(fileBytes, &blob)
defer fwpmFreeMemory0(&blob)
...

File: generate.go [hunk 2]
```diff
@@ -4,7 +4,6 @@
 
 package wf
 
-//go:generate go run generators/gen_guids.go includes/fwpmu.h zguids.go
```
Comment (danderson) [inline generate.go:7]: Don't remove this, it's still used (although you need a copy of the win10 SDK files manually placed at includes/... for it to work). It's fine to disable it while hacking, but it should stay in place.

**Review by danderson — CHANGES_REQUESTED** [review]

File: generate.go (hunk 2 above)
Comment (maisem) [reply generate.go:7]: Oh this wasn't supposed to be part of the pr

File: firewall.go (hunk 1 above)
Comment (maisem) [reply firewall.go:153]: Thanks for catching this!

File: firewall.go [hunk 3]
```diff
@@ -109,6 +109,54 @@ func (id *LayerID) IsZero() bool {
...
+// AppID returns the application ID associated with the provided file.
+func AppID(file string) (string, error) {
+	var a arena
+	defer a.Dispose()
+	fileBytes, _ := toBytesFromString(&a, file)
+	var appID *fwpByteBlob
+	if err := fwpmGetAppIdFromFileName0(fileBytes, &appID); err != nil {
+		return "", err
+	}
+	defer fwpmFreeMemory0((*struct{})(unsafe.Pointer(appID)))
```
Comment (danderson) [inline firewall.go:156]: The pointer passed to fwpmFreeMemory0 needs to be a pointer to the pointer you want to free, confusingly. IOW, it needs to also be &appID.

Learned the hard way while trying to figure out heisen-crashes early in wf dev :)

**Review by danderson — APPROVED** [review]

File: firewall.go (hunk 3 above)
Comment (maisem) [reply firewall.go]: Ah, yeah should have read the docs.

**Pull request merged**